06x-new-kafka-roller.md

```diff
@@ -0,0 +1,220 @@
+   - `RESTART_FIRST` - Nodes that have `NOT_READY` or `NOT_RUNNING` state in their contexts. The group will also include nodes that we cannot connect to via Admin API.
+   - `WAIT_FOR_LOG_RECOVERY` -  Nodes that have `RECOVERING` state.
+   - `RESTART` - Nodes that have non-empty list of reasons from the predicate function and have not been restarted yet (Context.numRestarts == 0).
+   - `MAYBE_RECONFIGURE` - Broker nodes (including combined nodes) that have an empty list of reasons and not been reconfigured yet (Context.numReconfigs == 0).
+   - `NOP` - Nodes that have been restarted or reconfigured at least once (Context.numRestarts > 0 || Context.numReconfigs > 0 ) and have either
+             `LEADING_ALL_PREFERRED` or `SERVING` state.
+
+4. Wait for nodes in `WAIT_FOR_LOG_RECOVERY` group to finish performing log recovery. 
+   - Wait for each node to have `SERVING` within the `postOperationalTimeoutMs`. 
+   - If the timeout is reached for a node and its `numAttempts` is greater than or equal to `maxAttempts`, throw `UnrestartableNodesException` with the log recovery progress (number of remaining logs and segments). Otherwise increment node's `numAttempts` and restart from step 3.
+
+5. Restart nodes in `RESTART_FIRST` category either one by one in the following order unless all nodes are combined
+and are in `NOT_RUNNING` state:
+   - Pure controller nodes
+   - Combined nodes
+   - Broker only nodes
+
+   If all controllers are combined and are in `NOT_RUNNING` state, restart all nodes in parallel and wait for them to have `SERVING`. Explained more in detail below.
+
+   Wait until the restarted node to have `SERVING` within `postOperationalTimeoutMs`. If the timeout is reached and the node's `numAttempts` is greater than or equal to `maxAttempts`, throw `TimeoutException`. Otherwise increment node's `numAttempts` and restart from step 3. 
+
+6. Further refine the broker nodes in `MAYBE_RECONFIGURE` group:
+   - Describe Kafka configurations for each node via Admin API and compare them against the desired configurations. This is essentially the same mechanism we use today for the current KafkaRoller.
+   - If a node has configuration changes and they can be dynamically updated, add the node into another group called `RECONFIGURE`.
+   - If a node has configuration changes but they cannot be dynamically updated, add nodes into the `RESTART` group.
+   - If a node has no configuration changes, put the node into the `NOP` group.
+
+7. Reconfigure each node in `RECONFIGURE` group:
+   - If `numReconfigs` of a node is greater than the configured `maxReconfigs`, add a restart reason to its context. Otherwise continue.
+   - Send `incrementalAlterConfig` request with its config updates.
+   - Transitions the node's state to `RECONFIGURED` and increment its `numReconfigs`.
+   - Wait for each node that got configurations updated until they have `SERVING` within the `postOperationalTimeoutMs`. 
```

Comment:
I know we've discussed this before, but the problem with this is that (unlike a restart) it's not obvious (when making observations) when the action (reconfiguration) has happened. There's a risk that we proceed based on a SERVING state that was based on stale information.

Reply:
This line was actually not correct. In our POC code, we wait for it to have `LEADING_ALL_PREFERRED` so it shouldn't be stale.